### PR TITLE
[glslang] Refine implicit array size interfaces.

### PR DESCRIPTION
**Pull request by ShchchowAMD**
Help to check builtin and other variables if across stages.

### CODE REVIEW

**Comment by greg-lunarg** [comment]
Can you please add tests that demonstrate the new functionality? Similar to #3066, although I think you will instead register in Link.FromFile.cpp.

**Comment by ShchchowAMD** [comment]
Thanks for reminding. I'll add test for this one and also future PRs.

**Comment by ShchchowAMD** [comment]
> Can you please add tests that demonstrate the new functionality? Similar to #3066, although I think you will instead register in Link.FromFile.cpp.

Add tests for link time error check.

File: glslang/MachineIndependent/linkValidate.cpp [hunk 1]
```diff
@@ -744,9 +744,19 @@ void TIntermediate::mergeLinkerObjects(TInfoSink& infoSink, TIntermSequence& lin
...
+                        symbol->getWritableType().updateImplicitArraySize(unitSymbol->getType().getImplicitArraySize());
+                    }
+                }
+                else if (symbol->getWritableType().isImplicitlySizedArray() && unitSymbol->getType().isSizedArray()) {
+                    if (symbol->getWritableType().getImplicitArraySize() > unitSymbol->getType().getOuterArraySize())
+                        error(infoSink, "Implicit size of unsized array aren't match same symbol among multiple shaders.");
+                }
+                else if (unitSymbol->getType().isImplicitlySizedArray() && symbol->getWritableType().isSizedArray()) {
+                    if (unitSymbol->getType().getImplicitArraySize() > symbol->getWritableType().getOuterArraySize())
+                        error(infoSink, "Implicit size of unsized array aren't match same symbol among multiple shaders.");
```
Comment (greg-lunarg) [inline glslang/MachineIndependent/linkValidate.cpp:759]: "aren't" -> "doesn't"

File: glslang/MachineIndependent/linkValidate.cpp [hunk 2]
```diff
@@ -744,9 +744,19 @@ void TIntermediate::mergeLinkerObjects(TInfoSink& infoSink, TIntermSequence& lin
                 if (! symbol->getQualifier().hasBinding() && unitSymbol->getQualifier().hasBinding())
                     symbol->getQualifier().layoutBinding = unitSymbol->getQualifier().layoutBinding;
 
-                // Similarly for location
-                if (!symbol->getQualifier().hasLocation() && unitSymbol->getQualifier().hasLocation()) {
-                    symbol->getQualifier().layoutLocation = unitSymbol->getQualifier().layoutLocation;
```
Comment (greg-lunarg) [inline glslang/MachineIndependent/linkValidate.cpp:749]: Did you mean to remove this? If so, why?

File: glslang/MachineIndependent/linkValidate.cpp [hunk 3]
```diff
@@ -759,6 +769,19 @@ void TIntermediate::mergeLinkerObjects(TInfoSink& infoSink, TIntermSequence& lin
...
+        // Check conflicts between preset primitives and sizes of I/O variables among multiple geometry shaders
+        if (language == EShLangGeometry && unitStage == EShLangGeometry)
+        {
+            TIntermSymbol* unitSymbol = unitLinkerObjects[unitLinkObj]->getAsSymbolNode();
+            if (unitSymbol->isArray() && unitSymbol->getQualifier().storage == EvqVaryingIn && unitSymbol->getQualifier().builtIn == EbvNone)
+                if ((unitSymbol->getArraySizes()->isImplicitlySized() &&
+                        unitSymbol->getArraySizes()->getImplicitSize() != TQualifier::mapGeometryToSize(getInputPrimitive())) ||
+                    (! unitSymbol->getArraySizes()->isImplicitlySized() &&
+                        unitSymbol->getArraySizes()->getDimSize(0) != TQualifier::mapGeometryToSize(getInputPrimitive())))
+                    error(infoSink, "Not all provided sizes match across all geometry shaders in the program");
```
[the remark this answers is not in the document]
Comment (ShchchowAMD) [reply glslang/MachineIndependent/linkValidate.cpp:782]: Yes, I'll fix it.

File: glslang/MachineIndependent/linkValidate.cpp (hunk 1 above)
Comment (ShchchowAMD) [reply glslang/MachineIndependent/linkValidate.cpp:759]: Thanks, will fix in next commit.

File: glslang/MachineIndependent/linkValidate.cpp (hunk 2 above)
Comment (ShchchowAMD) [reply glslang/MachineIndependent/linkValidate.cpp:749]: Sorry, this is a mistake via merging codes, I'll add this block back.

**Comment by ShchchowAMD** [comment]
Add a fix for corner case:
when a builtin variable is used across stages, set initialized implicit size to 0.

If stage A uses it by implicitly indexing but not in stage B, it should not report an error.

This is especially for vars in gl_PerVertex and other built-in blocks available across stages.

The main idea is to use implicitSize == 0 to determine whether a declared unsized array has been visited with an index.

For example:

VS:

...
void c(vec4 p, vec4 c[8], uint x)
{
	gl_ClipDistance[7] = mix(1.0, dot(p, c[7]), x(c, 7u));
}
...
void main() {
  ...
  c(...)
}


GS (CLIP not defined):

...
#if defined(CLIP)
void o(uint i)
{
	gl_ClipDistance[7] = gl_in[i].gl_ClipDistance[7];
}
#endif
...
void main(){
  ...
  #if defined(CLIP)
	  o(0u);
  #endif
  ...
}

**Comment by ShchchowAMD** [comment]
Test info should be changed.
Latest added commit for default set implicit sized check needs a test for it.
I'll add a new commit today.

File: glslang/MachineIndependent/linkValidate.cpp [hunk 4]
```diff
@@ -749,6 +749,21 @@ void TIntermediate::mergeLinkerObjects(TInfoSink& infoSink, TIntermSequence& lin
...
 
+                // Update implicit array sizes
+                if (symbol->getWritableType().isImplicitlySizedArray() && unitSymbol->getType().isImplicitlySizedArray()) {
+                    if (unitSymbol->getType().getImplicitArraySize() > symbol->getType().getImplicitArraySize()){
+                        symbol->getWritableType().updateImplicitArraySize(unitSymbol->getType().getImplicitArraySize());
+                    }
+                }
+                else if (symbol->getWritableType().isImplicitlySizedArray() && unitSymbol->getType().isSizedArray()) {
+                    if (symbol->getWritableType().getImplicitArraySize() > unitSymbol->getType().getOuterArraySize())
+                        error(infoSink, "Implicit size of unsized array aren't match same symbol among multiple shaders.");
```
Comment (greg-lunarg) [inline glslang/MachineIndependent/linkValidate.cpp:760]: aren't -> doesn't :)

Also in the .out file. Thanks!

**Review by greg-lunarg — CHANGES_REQUESTED** [review]
Just one small change that got missed. Thanks for your persistence!

File: glslang/MachineIndependent/linkValidate.cpp (hunk 4 above)
Comment (ShchchowAMD) [reply glslang/MachineIndependent/linkValidate.cpp]: Sorry, I forgot this one :)
New commit has fixed this error message.

**Review by greg-lunarg — APPROVED** [review]
LGTM. Thanks for the code!

**Pull request merged**